3.4_functional_programming.ipynb

```diff
@@ -425,40 +401,45 @@
     "val window = TriangularWindow\n",
     "\n",
     "// test our FIR\n",
-    "Driver(() => new MyFir(length, bitwidth, window)) {\n",
-    "  c => new PeekPokeTester(c) {\n",
+    "test(new MyFir(length, bitwidth, window)) { c =>\n",
     "    \n",
     "    // test data\n",
     "    val n = 100 // input length\n",
     "    val sine_freq = 10\n",
     "    val samp_freq = 100\n",
-    "    \n",
+    "\n",
     "    // sample data, scale to between 0 and 2^bitwidth\n",
     "    val max_value = pow(2, bitwidth)-1\n",
     "    val sine = (0 until n).map(i => (max_value/2 + max_value/2*sin(2*Pi*sine_freq/samp_freq*i)).toInt)\n",
     "    //println(s\"input = ${sine.toArray.deep.mkString(\", \")}\")\n",
-    "    \n",
+    "\n",
     "    // coefficients\n",
     "    val coeffs = window(length, bitwidth)\n",
     "    //println(s\"coeffs = ${coeffs.toArray.deep.mkString(\", \")}\")\n",
     "\n",
     "    // use breeze filter as golden model; need to reverse coefficients\n",
-    "    val expected = filter(DenseVector(sine.toArray), \n",
-    "                          FIRKernel1D(DenseVector(coeffs.reverse.toArray), 1.0, \"\"), \n",
-    "                          OptOverhang.None)\n",
-    "    //println(s\"exp_out = ${expected.toArray.deep.mkString(\", \")}\")\n",
+    "    val expected = filter(\n",
+    "        DenseVector(sine.toArray),\n",
+    "        FIRKernel1D(DenseVector(coeffs.reverse.toArray), 1.0, \"\"),\n",
+    "        OptOverhang.None\n",
+    "    )\n",
+    "    expected.toArray // seems to be necessary\n",
+    "    //println(s\"exp_out = ${expected.toArray.deep.mkString(\", \")}\") // this seems to be necessary\n",
     "\n",
     "    // push data through our FIR and check the result\n",
-    "    reset(5)\n",
+    "    c.reset.poke(true.B)\n",
+    "    c.clock.step(5)\n",
+    "    c.reset.poke(false.B)\n",
     "    for (i <- 0 until n) {\n",
```

Comment:
Could this be better written as a for loop to initialize, then a subsequent for loop that includes the expected checks? Possibly with the loop iterating over the expected value array?